Generate Certificate Request with predictable name

sathyanarays opened the pull request:
Modify certificaterequest controller to generate CertificateRequest with name as function of certificate name and revision.

<!--

Thanks for opening a pull request! Here are some tips to get everything merged smoothly:

1. Read our contributor guidelines: https://cert-manager.io/docs/contributing/

2. Make sure your commits are signed off: https://cert-manager.io/docs/contributing/sign-off/

3. If the PR is unfinished, raise it as a draft or prefix the title with "WIP:" so it's clear to everyone.

4. Be sure to allow edits from maintainers so it's easier for us to help: https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

-->

### Pull Request Motivation
As mentioned in #4956 , multiple CertificateRequest objects are created for a fixed certificate object and fixed revision. This happens when the CertificateRequest creation request from `requestmanager` controller takes more than 5 seconds to create the resource in the api server.

On timeout, the request is re-tried and a new CertificateRequest is created but with a different name. This results in multiple CertificateRequests for a given certificate name and given revision.

In this PR, this issue is being fixed by making the CertificateRequest name to be a function of certificate name and the revision. This makes sure, even if there are multiple CertificateRequest create request, they are requested with same name. Hence, this avoid multiple CertificateRequests for given cert and revision.

Other fix: #5485 
<!-- Explain the motivation behind this PR. If there's a related issue or PR, link to it here! -->

### Kind
bug
<!--

Pick a kind which best describes your PR from the following list:

	<cleanup | bug | feature | documentation | design | flake>

If you're unsure which is best or if you're not sure what we mean by "kind",
just ignore this section and a maintainer will fill it in for you!
-->

### Local testing
This change has been tested by simulating API server delay locally by the following diff:

```
diff --git a/pkg/controller/certificates/requestmanager/requestmanager_controller.go b/pkg/controller/certificates/requestmanager/requestmanager_controller.go
index 155580ec5..af87dd03c 100644
--- a/pkg/controller/certificates/requestmanager/requestmanager_controller.go
+++ b/pkg/controller/certificates/requestmanager/requestmanager_controller.go
@@ -395,13 +395,16 @@ func (c *controller) createNewCertificateRequest(ctx context.Context, crt *cmapi
 		},
 	}
 
-	cr, err = c.client.CertmanagerV1().CertificateRequests(cr.Namespace).Create(ctx, cr, metav1.CreateOptions{FieldManager: c.fieldManager})
-	if err != nil {
-		c.recorder.Eventf(crt, corev1.EventTypeWarning, reasonRequestFailed, "Failed to create CertificateRequest: "+err.Error())
-		return err
-	}
+	go func() {
+		time.Sleep(1 * time.Minute)
+		cr, err = c.client.CertmanagerV1().CertificateRequests(cr.Namespace).Create(ctx, cr, metav1.CreateOptions{FieldManager: c.fieldManager})
+		if err != nil {
+			c.recorder.Eventf(crt, corev1.EventTypeWarning, reasonRequestFailed, "Failed to create CertificateRequest: "+err.Error())
+		}
+
+		c.recorder.Eventf(crt, corev1.EventTypeNormal, reasonRequested, "Created new CertificateRequest resource %q", cr.Name)
+	}()
 
-	c.recorder.Eventf(crt, corev1.EventTypeNormal, reasonRequested, "Created new CertificateRequest resource %q", cr.Name)
 	if err := c.waitForCertificateRequestToExist(cr.Namespace, cr.Name); err != nil {
 		return fmt.Errorf("failed whilst waiting for CertificateRequest to exist - this may indicate an apiserver running slowly. Request will be retried. %w", err)
 	}

```

### Release Note

<!--

Should we mention this PR in release notes? If so, replace "NONE" with a line of text explaining what changed!

For more details, see: https://git.k8s.io/community/contributors/guide/release-notes.md

-->

```release-note
The feature to create certificate requests with the name being a function of certificate name and revision has been introduced under the feature flag "StableCertificateRequestName" and it is disabled by default. This helps to prevent the error "multiple CertificateRequests were found for the 'next' revision...". 
```

munnerz commented:
/ok-to-test
/hold

This is quite a sensitive area of code to change, so I don't think we should just make it like this without any feature gating etc. I'm also concerned about the side effects of this change - if I remember correctly, there are certain edge cases that can come up in our controllers that have _benefited_ from having a generated name, though I can't specifically recall them now.

If we do eventually change the default behaviour here, we may also need to communicate to our users that we are changing the naming format as in the past when we have changed names of things, we have found users have come to rely on the old format & built automation around it. I'm sure we _can_ do it, but it may require a heads up for people.

For now, can you add a featuregate for this functionality, something like `StableChallengeName`, which is off by default?

sathyanarays commented:
/retest

sathyanarays commented:
> /ok-to-test /hold
> 
> This is quite a sensitive area of code to change, so I don't think we should just make it like this without any feature gating etc. I'm also concerned about the side effects of this change - if I remember correctly, there are certain edge cases that can come up in our controllers that have _benefited_ from having a generated name, though I can't specifically recall them now.
> 
> If we do eventually change the default behaviour here, we may also need to communicate to our users that we are changing the naming format as in the past when we have changed names of things, we have found users have come to rely on the old format & built automation around it. I'm sure we _can_ do it, but it may require a heads up for people.
> 
> For now, can you add a featuregate for this functionality, something like `StableChallengeName`, which is off by default?

@munnerz , done!

AcidLeroy commented:
@sathyanarays & @munnerz I like this solution and would be happy to test out the feature gate. We run enough Kubernetes clusters to possibly weed out any disadvantages of not having the unique name when creating certificate requests, too. I spent a few minutes today digging through the issues/discussions on GitHub and on Slack, but I could not find any obvious specific references to the certificate request name needing to be generated. 

Outside the solution proposed here, could there be another way to check if there is another `certificateRequest` with the same revision already in flight before the controller attempts to create a new one? Maybe there is a way to query for certificate requests that match the given name and then look inside it for the revision? Or is the revision not available until after the request succeeds?

sathyanarays commented:
> @sathyanarays & @munnerz I like this solution and would be happy to test out the feature gate. We run enough Kubernetes clusters to possibly weed out any disadvantages of not having the unique name when creating certificate requests, too. I spent a few minutes today digging through the issues/discussions on GitHub and on Slack, but I could not find any obvious specific references to the certificate request name needing to be generated.
> 
> Outside the solution proposed here, could there be another way to check if there is another `certificateRequest` with the same revision already in flight before the controller attempts to create a new one? Maybe there is a way to query for certificate requests that match the given name and then look inside it for the revision? Or is the revision not available until after the request succeeds?

@AcidLeroy , in the same file, line 184-230 checks if the certificates for the same revision already exists. It creates new certificate request only if there is no existing certificate request for the same revision. But, these checks did not work as the inflight create requests are not showing up in these checks. Only, succeeded certificate requests are returned.

```
	requests, err := certificates.ListCertificateRequestsMatchingPredicates(c.certificateRequestLister.CertificateRequests(crt.Namespace), labels.Everything(), predicate.ResourceOwnedBy(crt))
	if err != nil {
		return err
	}

	// delete any existing CertificateRequest resources that do not have a
	// revision annotation
	if requests, err = c.deleteRequestsWithoutRevision(ctx, requests...); err != nil {
		return err
	}

	currentCertificateRevision := 0
	if crt.Status.Revision != nil {
		currentCertificateRevision = *crt.Status.Revision
	}
	nextRevision := currentCertificateRevision + 1

	requests, err = requestsWithRevision(requests, nextRevision)
	if err != nil {
		return err
	}

	requests, err = c.deleteRequestsNotMatchingSpec(ctx, crt, pk.Public(), requests...)
	if err != nil {
		return err
	}

	requests, err = c.deleteCurrentFailedRequests(ctx, crt, requests...)
	if err != nil {
		return err
	}

	if len(requests) > 1 {
		// TODO: we should handle this case better, but for now do nothing to
		//  avoid getting into loops where we keep creating multiple requests
		//  and deleting them again.
		log.V(logf.ErrorLevel).Info("Multiple matching CertificateRequest resources exist, delete one of them. This is likely an error and should be reported on the issue tracker!")
		return nil
	}

	if len(requests) == 1 {
		// Nothing to do as we've already verified that the CertificateRequest
		// is up to date above.
		return nil
	}

	return c.createNewCertificateRequest(ctx, crt, pk, nextRevision, nextPrivateKeySecret.Name)
```

The problem here is that the certificate request create operation is **neither transactional nor idempotent**. 

The operation `c.createNewCertificateRequest(ctx, crt, pk, nextRevision, nextPrivateKeySecret.Name)` is not transactional because successful call of this function does not mean successful creation of certificate request resource. Also, failure of this function call may not always mean that the certificate resource will not get created.

The best we could do here is to make the create operations idempotent. This is done by creating predictable names for the certificate requests in this PR. There may be other ways to bring in the idempotency. 

I could think of one more way to bring in the same idempotency property. This can be done by storing the name of the certificate request that got generated in the status section of the certificate request. And we have to modify the code such that any concurrent requests should check the certificate request name in the certificate resource before proceeding with creation. Though this approach will work, there will be more changes to the core of cert-manager (we have to change the Certificate CRD for example) and we might need to think thoroughly about backward compatibility. Hence, I chose this simpler approach.

AcidLeroy commented:
@sathyanarays Thank you for the outline, well put!

> I could think of one more way to bring in the same idempotency property. This can be done by storing the name of the certificate request that got generated in the status section of the certificate request. And we have to modify the code such that any concurrent requests should check the certificate request name in the certificate resource before proceeding with creation. Though this approach will work, there will be more changes to the core of cert-manager (we have to change the Certificate CRD for example) and we might need to think thoroughly about backward compatibility. Hence, I chose this simpler approach.

How would we store the status of the certificate request? If the apiserver is running slow, wouldn't we potentially have the same timeout issue, or would the proposed status update not need to go through the apiserver? If @munnerz is worried about unintended consequences, maybe the more complex approach would be better in this case? He made it sound like there were a few use cases that relied on the generated names. If we can solve the problem without fundamentally changing the naming scheme of the certificate requests, should we perhaps prefer that one? Anyway, I'm more just thinking outloud here.

sathyanarays commented:
> @sathyanarays Thank you for the outline, well put!
> 
> > I could think of one more way to bring in the same idempotency property. This can be done by storing the name of the certificate request that got generated in the status section of the certificate request. And we have to modify the code such that any concurrent requests should check the certificate request name in the certificate resource before proceeding with creation. Though this approach will work, there will be more changes to the core of cert-manager (we have to change the Certificate CRD for example) and we might need to think thoroughly about backward compatibility. Hence, I chose this simpler approach.
> 
> How would we store the status of the certificate request? If the apiserver is running slow, wouldn't we potentially have the same timeout issue, or would the proposed status update not need to go through the apiserver? If @munnerz is worried about unintended consequences, maybe the more complex approach would be better in this case? He made it sound like there were a few use cases that relied on the generated names. If we can solve the problem without fundamentally changing the naming scheme of the certificate requests, should we perhaps prefer that one? Anyway, I'm more just thinking outloud here.

@AcidLeroy , in the alternate approach I mentioned, we are going to generate a name for certificate request and store it in the certificate resource and save the certificate resource. If there are multiple outstanding updates on a certificate resource (due to slow API server), exactly one will succeed. We will use the name from the succeeded update to generate the certificate resources.

I understand that it will be slightly difficult to visualise things. What about I make the actual changes, test it out and come up with a draft PR?

munnerz commented:
To clarify, the original bug here is not anything to do with create requests not succeeding, nor is it because only completed certificate requests are being returned.

The bug here comes about because everything in a controller is driven by the cache, which is a reflection of the apiserver's contents. When the first 'create' call is submitted, a 200 (or 201) response is returned and that object HAS been created. The bug we have hit though, is that it's taking too long (>5s in this case), for our local cache to _also_ observe the recently created new object due to some kind of network latency, connection drop, or apiserver instability.

The object DOES already exist, it's just our own cached copy of CRs that is stored in memory is not up to date enough yet to observe it having been created.

That means that when we re-reconcile the object, we check to see if a CR for this revision already exists once again, but as our cache is out of date the controller _still_ believes that one does not exist, so it creates another.

By making the name predictable/static, we can avoid this happening, because when we go to submit that second create call in the above example, an AlreadyExists error will be returned from the server.

Hope that clears things up a bit 😊

munnerz commented:
@sathyanarays we cannot store the name of the CR on the Certificate object for exactly the same reason - we once again may not observe our own 'update' operation on the Certificate before we go and re-reconcile the object, creating the same problem again 😅

The only way to make that sort of strategy work is to set the name *before* firing a Create event, as we could lean on optimistic concurrency with the resourceVersion to protect us from conflicts. However that is effectively what we are looking to do with the revision parameter already - we are persisting an ID number that helps us uniquely identify the CertificateRequest associated with a certificate/revision.

sathyanarays commented:
> The only way to make that sort of strategy work is to set the name before firing a Create event

@munnerz , that is part of the alternate approach that I described. But, somehow missed to express that verbally.

> However that is effectively what we are looking to do with the revision parameter already - we are persisting an ID number that helps us uniquely identify the CertificateRequest associated with a certificate/revision.

Exactly. The approach in the current PR effectively brings the same behavior described in the alternate approach!

sathyanarays commented:
> To clarify, the original bug here is not anything to do with create requests not succeeding, nor is it because only completed certificate requests are being returned.
> 
> The bug here comes about because everything in a controller is driven by the cache, which is a reflection of the apiserver's contents. When the first 'create' call is submitted, a 200 (or 201) response is returned and that object HAS been created. The bug we have hit though, is that it's taking too long (>5s in this case), for our local cache to _also_ observe the recently created new object due to some kind of network latency, connection drop, or apiserver instability.
> 
> The object DOES already exist, it's just our own cached copy of CRs that is stored in memory is not up to date enough yet to observe it having been created.
> 
> That means that when we re-reconcile the object, we check to see if a CR for this revision already exists once again, but as our cache is out of date the controller _still_ believes that one does not exist, so it creates another.
> 
> By making the name predictable/static, we can avoid this happening, because when we go to submit that second create call in the above example, an AlreadyExists error will be returned from the server.
> 
> Hope that clears things up a bit 😊

@munnerz , Thanks for clarifying. This helps to understand the problem better!

munnerz commented:
@sathyanarays can you also change where we do the `wait` call so that we don't do this waiting *if* this feature gate is enabled? The point here being that it's no longer required for us to wait if the name is stable/predictable (as we will rely on IsNotFound errors instead).

irbekrm commented on internal/controller/feature/features.go:
```diff
@@ -58,6 +58,12 @@ const (
 	// See https://github.com/cert-manager/cert-manager/issues/3203 and https://github.com/cert-manager/cert-manager/issues/4424 for context.
 	LiteralCertificateSubject featuregate.Feature = "LiteralCertificateSubject"
+
+	// Alpha: v1.9.1
```
I think this could be 1.10 or a 1.11 alpha release if it does not make it into 1.10
I don't think we would add a new feature in a patch release

sathyanarays replied:
Changed it to 1.10 for now.

irbekrm commented:
Is it possible to add some tests for this? Also, we enable all feature gates for e2e test by default [here](https://github.com/cert-manager/cert-manager/blob/master/make/e2e-setup.mk#L161) and [here](https://github.com/cert-manager/cert-manager/blob/master/make/e2e-setup.mk#L171)

sathyanarays commented:
> @sathyanarays can you also change where we do the `wait` call so that we don't do this waiting _if_ this feature gate is enabled? The point here being that it's no longer required for us to wait if the name is stable/predictable (as we will rely on IsNotFound errors instead).

@munnerz , done!

sathyanarays commented:
> Is it possible to add some tests for this? Also, we enable all feature gates for e2e test by default [here](https://github.com/cert-manager/cert-manager/blob/master/make/e2e-setup.mk#L161) and [here](https://github.com/cert-manager/cert-manager/blob/master/make/e2e-setup.mk#L171)

@irbekrm , done!

munnerz reviewed (commented):
Added a few very minor comments, but otherwise this looks good to me! 😄 going to add `/kind feature` to this - when we come to flip this to default to true, I think it may be worth classifying that PR as a bugfix.. not sure 🤷 

/kind feature
/approve
(leaving lgtm for after comments are fixed up)

sathyanarays commented:
> Added a few very minor comments, but otherwise this looks good to me! 😄 going to add `/kind feature` to this - when we come to flip this to default to true, I think it may be worth classifying that PR as a bugfix.. not sure 🤷
> 
> /kind feature /approve (leaving lgtm for after comments are fixed up)

@munnerz , all resolved!

munnerz commented on pkg/controller/certificates/requestmanager/requestmanager_controller.go:
```diff
@@ -395,15 +397,22 @@ func (c *controller) createNewCertificateRequest(ctx context.Context, crt *cmapi
 	c.recorder.Eventf(crt, corev1.EventTypeNormal, reasonRequested, "Created new CertificateRequest resource %q", cr.Name)
-	if err := c.waitForCertificateRequestToExist(cr.Namespace, cr.Name); err != nil {
-		return fmt.Errorf("failed whilst waiting for CertificateRequest to exist - this may indicate an apiserver running slowly. Request will be retried. %w", err)
+	if !utilfeature.DefaultFeatureGate.Enabled(feature.StableCertificateRequestName) {
```
Sorry, just caught this.. perhaps add a quick comment here explaining why we don't call `waitForCertificateRequestToExist` if the feature is enabled 👀

```
// If the StableCertificateRequestName feature gate is enabled, skip waiting for our informer cache/lister to
// observe the creation event and instead rely on an AlreadyExists error being returned if we do attempt a
// CREATE for the same CertificateRequest name again early.
```

Also, could you switch this `if` statement to assert the _positive_, e.g:

```
if utilfeature.DefaultFeatureGate.Enabled(feature.StableCertificateRequestName) {
    return nil
}

// ... existing `waitForCertificateRequestToExist` code here
```

Again, just trying to make it very clear and explicit when anyone looks at this code which behaviour is being gated :)

sathyanarays replied:
Done!

munnerz replied:
Thanks 🙌 that makes this changeset look very neat/clear that we aren't touching any existing/un-gated code paths 🥳

munnerz commented:
Thanks for the prompt response! I have just one more comment which I should have spotted earlier.. after that though, lgtm and let's get this out for v1.10 🥳

munnerz commented:
Looking forward to seeing some of the results of this feature gate being enabled! 🥳 thanks very much for getting this one over the line! 🚀 

/lgtm
/hold cancel

munnerz commented:
hmm, looks like we don't really have an appropriate `area` label for this 👀 

/remove-area testing

munnerz commented:
Would you mind updating the release note now that this is feature gated, just to make it clear that we've added a new (gated) feature? 😄

sathyanarays commented:
> Would you mind updating the release note now that this is feature gated, just to make it clear that we've added a new (gated) feature? 😄

@munnerz , updated. Just let me know if you have a better one!
